### PR TITLE
expression: add nullable() method to check whether an expression can return null

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

This PR introduce a new method `Expression.Nullable()` to check whether an expression can return null values. It is truly useful for query optimization. 

### What is changed and how it works?

The basic interface of `Nullable()` is:
```golang
type Expression interface {
  ...
  Nullable(notNullCols []*Column) bool
}
```

If an expression can return null value, `Nullable()` returns true, otherwise returns false.

There are three kinds of Expression: Constant, Column, ScalarFunction.
* `Contant.Nullable(...)` returns true if its value is null.
* `Column.Nullable(notNullCols)` returns true if the column is not in `notNullCols`.
* `ScalarFunction.Nullable(notNullCols)` depends on its builtinFunc's `nullable()` value.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

File: expression/builtin_arithmetic_vec.go [hunk 1]
```diff
@@ -110,6 +118,10 @@ func (b *builtinArithmeticModIntSig) vectorized() bool {
 	return true
 }
 
+func (b *builtinArithmeticModIntSig) nullable(notNullCols []*Column) bool {
+	return b.existsNullableChild(notNullCols)
```
Comment (kennytm) [inline expression/builtin_arithmetic_vec.go]: what about `5 MOD 0`

File: expression/builtin_arithmetic_vec.go [hunk 2]
```diff
@@ -648,6 +688,10 @@ func (b *builtinArithmeticIntDivideDecimalSig) vectorized() bool {
 	return true
 }
 
+func (b *builtinArithmeticIntDivideDecimalSig) nullable(notNullCols []*Column) bool {
+	return b.existsNullableChild(notNullCols)
```
Comment (kennytm) [inline expression/builtin_arithmetic_vec.go]: `5 / 0.0` ?

File: expression/builtin_arithmetic_vec.go [hunk 3]
```diff
@@ -736,6 +780,10 @@ func (b *builtinArithmeticMultiplyIntSig) vectorized() bool {
 	return true
 }
 
+func (b *builtinArithmeticMultiplyIntSig) nullable(notNullCols []*Column) bool {
+	return true
```
Comment (kennytm) [inline expression/builtin_arithmetic_vec.go]: if overflow is nullable here, why is `builtinArithmeticPlusIntSig` considered non-nullable (and why does this function call `result.SetNull()` on error while all other doesn't)?

File: expression/builtin_arithmetic_vec.go [hunk 4]
```diff
@@ -817,6 +865,10 @@ func (b *builtinArithmeticIntDivideIntSig) vectorized() bool {
 	return false
 }
 
+func (b *builtinArithmeticIntDivideIntSig) nullable(notNullCols []*Column) bool {
+	return b.existsNullableChild(notNullCols)
```
Comment (kennytm) [inline expression/builtin_arithmetic_vec.go]: `5 / 0`

File: expression/builtin_arithmetic_vec.go (hunk 3 above)
Comment (francis0407) [reply expression/builtin_arithmetic_vec.go]: I think it is because of the compatibility with mysql.

File: expression/constant.go [hunk 5]
```diff
@@ -62,6 +62,12 @@ type Constant struct {
 	hashcode    []byte
 }
 
+// Nullable implements Expression Nullable interface.
+// It returns whether this Constant is null.
+func (c *Constant) Nullable(notNullCols []*Column) bool {
+	return c.Value.IsNull()
```
Comment (qw4990) [inline expression/constant.go]: Should we consider `c.DeferredExpr` here?

**Comment by zz-jason** [comment]
can we close this PR since it hasn't been updated for nearly 1 year?

**Comment by zz-jason** [comment]
I'm going to close this PR due to it's stale, feel free to reopen it if you are going to update this PR in the future.

**Pull request closed without merging**